FWCore/Framework/interface/ESProducer.h

```diff
@@ -193,16 +233,32 @@ namespace edm {
         method in order to do the registration with the EventSetup
     */
     template <typename T, typename TReturn, typename TRecord, typename TArg>
-    ESConsumesCollectorT<TRecord> setWhatProduced(T* iThis,
-                                                  TReturn (T ::*iMethod)(const TRecord&),
-                                                  const TArg& iDec,
-                                                  const es::Label& iLabel = {}) {
+    auto setWhatProduced(T* iThis,
+                         TReturn (T ::*iMethod)(const TRecord&),
+                         const TArg& iDec,
+                         const es::Label& iLabel = {}) {
+      return setWhatProduced<TReturn, TRecord>(
+          [iThis, iMethod](TRecord const& iRecord) { return (iThis->*iMethod)(iRecord); },
+          createDecoratorFrom(iThis, static_cast<const TRecord*>(nullptr), iDec),
+          iLabel);
+    }
+
+    template <typename TFunc>
+    auto setWhatProduced(TFunc&& func, const es::Label& iLabel = {}) {
```

Comment:
For now I've intentionally omitted the decorator functionality for the overload taking the lambda (functor really).

Reply:
Do you plan to ever add it? Or do you think the lambda can do that work now as well?

Reply:
I added a comment following our discussion (i.e. intentionally left out, not sure if really needed, can be added later)